models/db/common.go

```diff
@@ -20,3 +28,360 @@ func BuildCaseInsensitiveLike(key, value string) builder.Cond {
 	}
 	return builder.Like{"UPPER(" + key + ")", strings.ToUpper(value)}
 }
+
+// InsertOnConflictDoNothing will attempt to insert the provided bean but if there is a conflict it will not error out
+// This function will update the ID of the provided bean if there is an insertion
+// This does not do all of the conversions that xorm would do automatically but it does quite a number of them
+// once xorm has a working InsertOnConflictDoNothing this function could be removed.
+func InsertOnConflictDoNothing(ctx context.Context, bean interface{}) (bool, error) {
+	e := GetEngine(ctx)
+
+	tableName := x.TableName(bean, true)
+	table, err := x.TableInfo(bean)
+	if err != nil {
+		return false, err
+	}
+
+	autoIncrCol := table.AutoIncrColumn()
+
+	columns := table.Columns()
+
+	colNames, values, zeroedColNames, zeroedValues, err := getColNamesAndValuesFromBean(bean, columns)
+	if err != nil {
+		return false, err
+	}
+
+	if len(colNames) == 0 {
+		return false, fmt.Errorf("provided bean to insert has all empty values")
+	}
+
+	// MSSQL needs to separately pass in the columns with the unique constraint and we need to
+	// include empty columns which are in the constraint in the insert for other dbs
+	uniqueCols, uniqueValues, colNames, values := addInUniqueCols(colNames, values, zeroedColNames, zeroedValues, table)
+	if len(uniqueCols) == 0 {
+		return false, fmt.Errorf("provided bean has no unique constraints")
+	}
+
+	var insertArgs []any
+
+	switch {
+	case setting.Database.UseSQLite3 || setting.Database.UsePostgreSQL || setting.Database.UseMySQL:
+		insertArgs = generateInsertNoConflictSQLAndArgs(tableName, colNames, values, autoIncrCol)
+	case setting.Database.UseMSSQL:
+		insertArgs = generateInsertNoConflictSQLAndArgsForMSSQL(tableName, colNames, values, uniqueCols, uniqueValues, autoIncrCol)
+	default:
+		return false, fmt.Errorf("database type not supported")
+	}
+
+	if autoIncrCol != nil && (setting.Database.UsePostgreSQL || setting.Database.UseMSSQL) {
+		// Postgres and MSSQL do not use the LastInsertID mechanism
+		// Therefore use query rather than exec and read the last provided ID back in
+
+		res, err := e.Query(insertArgs...)
+		if err != nil {
+			return false, fmt.Errorf("error in query: %s, %w", insertArgs[0], err)
+		}
+		if len(res) == 0 {
+			// this implies there was a conflict
+			return false, nil
+		}
+
+		aiValue, err := table.AutoIncrColumn().ValueOf(bean)
+		if err != nil {
+			log.Error("unable to get value for autoincrcol of %#v %v", bean, err)
+		}
+
+		if aiValue == nil || !aiValue.IsValid() || !aiValue.CanSet() {
+			return true, nil
+		}
+
+		id := res[0][autoIncrCol.Name]
+		err = convert.AssignValue(*aiValue, id)
+		if err != nil {
+			return true, fmt.Errorf("error in assignvalue %v %v %w", id, res, err)
+		}
+		return true, nil
+	}
+
+	res, err := e.Exec(values...)
```

Comment:
Does it need to be `Exec(insertArgs...)` here? Just a guess because there is a 
`Query(insertArgs...)` above

Reply:
yup mistake when I changed things